tyrian/js/src/main/scala/tyrian/Sub.scala

```diff
@@ -20,131 +23,181 @@ import scala.scalajs.js
-    * @tparam Msg
-    *   Type of message
-    */
-  def emit[Msg](msg: Msg): Sub[Msg] =
+  /** A subscription that emits a msg once. Identical to timeout with a duration of 0. */
+  def emit[F[_]: Sync, Msg](msg: Msg): Sub[F, Msg] =
     timeout(FiniteDuration(0, TimeUnit.MILLISECONDS), msg, msg.toString)
 
-  /** @return
-    *   A subscription that notifies its subscribers with `msg` after `duration`.
-    * @param duration
-    *   Duration of the timeout
-    * @param msg
-    *   Message produced by the timeout
-    * @param id
-    *   Globally unique identifier for this subscription
-    * @tparam Msg
-    *   Type of message
-    */
-  def timeout[Msg](duration: FiniteDuration, msg: Msg, id: String): Sub[Msg] =
-    ofTotalObservable[Msg](
-      id,
-      { observer =>
-        val handle =
-          dom.window.setTimeout(Functions.fun0(() => observer.onNext(msg)), duration.toMillis.toDouble)
-        () => dom.window.clearTimeout(handle)
+  /** A subscription that produces a `msg` after a `duration`. */
+  def timeout[F[_]: Sync, Msg](duration: FiniteDuration, msg: Msg, id: String): Sub[F, Msg] =
+    val acquire: F[(Either[Throwable, Msg] => Unit) => Int] =
+      Sync[F].delay { (callback: Either[Throwable, Msg] => Unit) =>
+        dom.window.setTimeout(
+          Functions.fun0(() => callback(Right(msg))),
+          duration.toMillis.toDouble
+        )
       }
-    )
-
-  def every(interval: FiniteDuration, id: String): Sub[js.Date] =
-    ofTotalObservable[js.Date](
-      id,
-      { observer =>
-        val handle =
-          dom.window.setInterval(Functions.fun0(() => observer.onNext(new js.Date())), interval.toMillis.toDouble)
-        () => dom.window.clearInterval(handle)
+
+    val release: F[Int => F[Option[F[Unit]]]] =
+      Sync[F].delay { (handle: Int) =>
+        Sync[F].delay {
+          Option(Sync[F].delay(dom.window.clearTimeout(handle)))
+        }
       }
-    )
 
-  def fromEvent[A, Msg](name: String, target: EventTarget)(extract: A => Option[Msg]): Sub[Msg] =
-    Sub.ofTotalObservable[Msg](
-      name + target.hashCode,
-      { observer =>
+    val task =
+      for {
+        a <- acquire
+        r <- release
+      } yield a andThen r
+
+    Observe(id, task)
+
+  /** A subscription that produces a `msg` after a `duration`. */
+  def timeout[F[_]: Sync, Msg](duration: FiniteDuration, msg: Msg): Sub[F, Msg] =
+    timeout(duration, msg, "[tyrian-sub-every] " + duration.toString + msg.toString)
+
+  /** A subscription that repeatedly produces a `msg` based on an `interval`. */
+  def every[F[_]: Sync](interval: FiniteDuration, id: String): Sub[F, js.Date] =
+    Sub.make[F, js.Date, Int](id) { callback =>
+      Sync[F].delay {
+        dom.window.setInterval(
+          Functions.fun0(() => callback(Right(new js.Date()))),
+          interval.toMillis.toDouble
+        )
+      }
+    } { handle =>
+      Sync[F].delay(dom.window.clearTimeout(handle))
+    }
```

Comment:
Side note: if possible, I'd still consider using `Temporal` here instead of `Sync`. I have no idea if it's important for frontend, but the Cats Effect test runtime is capable of mocking time, but only if you use the CE clock/time primitives rather than the raw APIs directly. For example, this lets you run tests with delays and stuff without actually having to wait for that time to pass, since it can instantly advance the clock, which is pretty cool.

https://typelevel.org/cats-effect/docs/core/test-runtime#mocking-time

---

Also, while I'm here I realize that I totally goofed the snippet I gave before, when I said this you could express `every` like this:
```scala
Temporal[F].sleep(interval) >> Temporal[F].realTimeDate
```
This would only run once and not _every_ interval 🤦 sorry about that. You could wrap a `repeatForever` around that, but then all that stuff about `join`ing the fiber is all wrong.

Without `Stream` I guess you _do_ need some sort of callback here. So, now I appreciate your design a lot better :)

Reply:
I had fun nosing around `Temporal` for a while there, but you're right, we're back to needing streams.

Interesting thing here is that you're unlikely to test that. Someone creates a `Sub` that emits a `Msg` after every elapsed interval. The `Msg` is handled in the `update` function. You'd test the update function, but you wouldn't bother testing the `Sub.every` - there's nothing to go wrong.

Of course, if someone wanted to do something fancy in their own custom `Sub`, they could certainly use a `Temporal` instance there if it helped them.

So at least for now, I think I'll leave it as-is.